### ISSUE TITLE
Flake is not backward compatible with usual non-flake commands

### ISSUE THREAD

**Issue by tobiasBora** [NixOS/nix#4024]
**Describe the bug**

Since I installed flake's nix, I saw some command changing in a quite disturbing way. For example `nix search mypackage` does not work anymore, I need to type `nix search nixpkgs mypackage` to make it work... and still now it takes quite some time to evaluate while before it was nearly instantaneous. Then, the `nix eval` function also changed; before I could write:

```
nixpkgs=$(nix eval --raw '(builtins.fetchTarball {
    url = https://github.com/NixOS/nixpkgs/archive/14006b724f3.tar.gz;
    sha256 = "07hfbilyh818pigfn342v2r05n8061wpjaf1m4h291lf6ydjagis";
  })')
```

but now I need to add the `--expr` tag. Since the behavior of the flake nix is quite different from the behaviour of the usual nix, don't you think that it would make sense to change the cli syntax to make sure that it does not interfere with old nix commands? I'm afraid that people will be lost when nix flake will become the "official" nix, if all the tutorials available start to give strange errors. And it also risks to break deployment scripts (like it did with mine).

**Comment by edolstra** [comment]
Yes, we changed some interfaces. The `nix` command is experimental, but this was not communicated very well. (E.g. `nix --help` notes that it's experimental.) So we now require `experimental-features` to be set to `nix-command` to be able to use it.

`nix search` was definitely not instantaneous before, but it had an evaluation cache (but so does the current implementation).

**Comment by tobiasBora** [comment]
Ok I see thanks. Also, even if people enable `experimental-features`, a strong warning that specifies that the commands are likely to change soon may be interesting.

Concerning the search, on my system the old `nix search` takes 0.3s to evaluate with the cache, while the cached `nix search nixpkgs` is about 5x slower (it writes "evaluate package-XXX").. and most annoying I think that the cache is recreated quite regularly (I've the feeling that ti's recreated at least once per day, so at least once per day a search takes ~45s but I need to check):

```
$ time nix search firefox-esr-68
warning: using cached results; pass '-u' to update the cache
* nixpkgs.firefox-esr-68 (firefox-68.9.0esr)
  A web browser built from Firefox Extended Support Release source tree (with plugins: )

* nixpkgs.firefox-esr-68-unwrapped (firefox-esr-unwrapped-68.9.0esr)
  A web browser built from Firefox Extended Support Release source tree

* nixpkgs.firefoxPackages.firefox-esr-68 (firefox-esr-unwrapped-68.9.0esr)
  A web browser built from Firefox Extended Support Release source tree


real    0m0,330s
user    0m0,284s
sys     0m0,047s

$ time nix search nixpkgs firefox-esr-68
error: --- Error --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- nix
no results for the given search term(s)!

real    0m43,634s
user    0m24,945s
sys     0m9,915s

[leo@bestos:~/Documents/Cours/2019-2022_-_These/Work_in_progress/2020_04_-_Code_projet_Antoine]$ time nix search nixpkgs firefox-esr
* legacyPackages.x86_64-linux.firefox-esr (78.2.0esr)
  A web browser built from Firefox Extended Support Release source tree (with plugins: )

* legacyPackages.x86_64-linux.firefox-esr-78 (78.2.0esr)
  A web browser built from Firefox Extended Support Release source tree (with plugins: )

* legacyPackages.x86_64-linux.firefox-esr-78-unwrapped (78.2.0esr)
  A web browser built from Firefox Extended Support Release source tree

* legacyPackages.x86_64-linux.firefox-esr-wrapper (78.2.0esr)
  A web browser built from Firefox Extended Support Release source tree (with plugins: )

* legacyPackages.x86_64-linux.firefoxPackages.firefox-esr-78 (78.2.0esr)
  A web browser built from Firefox Extended Support Release source tree

real    0m1,556s
user    0m1,359s
sys     0m0,113s
```

Anyway, I guess this is not really linked to the current issue, thanks for your answers. I'm just curious to know if it's normal that `nix search nixpkgs` clears its cache that often, and if not I'll investigate and eventually fill another issue if I can confirm that search clears its cache often.

**Comment by edolstra** [comment]
The new `nix search` cache is never cleared, but the `nixpkgs` flake can be updated several times a day, forcing `nix search` to update the cache. If you don't want that, you could pin the `nixpkgs` flake (via `nix flake pin`, or in your NixOS config, see "Pinning nixpkgs" in https://www.tweag.io/blog/2020-07-31-nixos-flakes/).

BTW, I'm not seeing such a major speed difference. On my laptop, the old search (for `firefox`) takes 10.4s on a cold cache and 0.3s on a hot cache. The new one takes 9.4s and 0.6s respectively. Make sure you're using the same nixpkgs though - `firefox-esr-68` exists in 20.03 but not on master so that might explain some of the difference.

The new cache uses SQLite. It's possible it's not performing well in some configurations...

**Comment by tobiasBora** [comment]
Oh I see, thanks!

One more question, after I stop bothering you: is there a stable way to evaluate an expression on nix? For example, I'd like to have a way to compile the following expression that does not depend on the nix flavor (flake/non flake):

```
nix eval --raw '(builtins.fetchTarball {
     url = https://github.com/NixOS/nixpkgs/archive/14006b724f3.tar.gz;
     sha256 = "07hfbilyh818pigfn342v2r05n8061wpjaf1m4h291lf6ydjagis";
   })
```

**Comment by edolstra** [comment]
The stable way is `nix-instantiate --eval --expr`.

**Comment by tobiasBora** [comment]
Ok thanks. But there is no `--raw` option to remove the trailing quotes `"..."`?

Exple:
```
$ nix-instantiate --eval -E 'builtins.fetchTarball {url = https://github.com/NixOS/nixpkgs/archive/925ae0dee63.tar.gz; sha256 = "1g3kkwyma23lkszdvgb4dn91g35b082k55ys8azc7j4s6vxpzmaw"; }'
"/nix/store/m59d3v89nhp9207hgvf6v21wxq7lhq64-source"
```

**Comment by tobiasBora** [comment]
Sorry to continue on this issue, but I tried to pin as you proposed nixpkgs in flake:

> The new `nix search` cache is never cleared, but the `nixpkgs` flake can be updated several times a day, forcing `nix search` to update the cache. If you don't want that, you could pin the `nixpkgs` flake (via `nix flake pin`, or in your NixOS config, see "Pinning nixpkgs" in https://www.tweag.io/blog/2020-07-31-nixos-flakes/).

However both option fails:

```
$ nix flake pin
[...] 'pin' is not a recognised command [...] 
$ nix flake lock nixpkgs
[...] 'list' is not a recognised command [...]
```
and if I use `registry.nixpkgs.flake = nixpkgs;`, then I get `undefined variable 'nixpkgs'`, I guess because my `configuration.nix` file is not yet converted to flake. Is there a solution if I don't want to use flake for my `configuration.nix`?

**Comment by edolstra** [comment]
Sorry, `nix flake pin` got renamed to `nix registry pin`.

> I guess because my configuration.nix file is not yet converted to flake

Right, `nixpkgs` here refers to the `nixpkgs` input of the flake.

> Is there a solution if I don't want to use flake for my configuration.nix?

You might be able to do something like `registry.nixpkgs.flake = builtins.fetchGit { ... };`.

**Comment by tobiasBora** [comment]
Ok, thanks a lot, `nix registry pin` seems to work (at least it does not give any error, we will see if the caching mechanism works or not)!

**Comment by milahu** [comment]
>> Flake is not backward compatible with usual non-flake commands
>
> Yes, we changed some interfaces.

not nice : /

> But there is no `--raw` option to remove the trailing quotes `"..."`?

nope. default output format is astree (abstract syntax tree)
other formats are json and xml

```
nix-instantiate --eval -E 'builtins.fetchTarball { ... }' --json  | jq --raw-output

/nix/store/m59d3v89nhp9207hgvf6v21wxq7lhq64-source
```

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
